File: projects/11-water-treatment-utility/fog/aggregation.js
"use strict";

// Pure window-aggregate math, kept separate from ledger.js's buffering
// mechanics so it can be unit tested against plain arrays of readings
// without touching the ledger/grouping plumbing at all. Runs once per
// group at flush time, over whatever readings groupByKey() handed back --
// never incrementally maintained during ingest.
function summarizeWindow(sensorType, siteId, unit, readings, windowStart, windowEnd) {
  const values = readings.map((r) => r.value);
  const sum = values.reduce((acc, v) => acc + v, 0);
  return {
    sensor_type: sensorType,
    site_id: siteId,
    unit,
    window_start: windowStart,
    window_end: windowEnd,
    count: values.length,
    min: Math.min(...values),
    max: Math.max(...values),
    avg: Math.round((sum / values.length) * 1000) / 1000,
    latest: readings[readings.length - 1].value,
  };
}

module.exports = { summarizeWindow };

File: projects/23-marine-vessel-monitoring/fog/aggregation.py
from statistics import fmean


def aggregate(sensor_type, site_id, unit, readings, window_start, window_end):
    """Reduce one window's buffered readings for a (sensor_type, site_id)
    vessel pair into a single summary record.

    latest is the last reading in arrival order, not the max-timestamp
    reading -- sensors dispatch batches in order, so arrival order already
    reflects chronological order without parsing every ts.
    """
    values = [r["value"] for r in readings]
    return {
        "sensor_type": sensor_type,
        "site_id": site_id,
        "unit": unit,
        "window_start": window_start,
        "window_end": window_end,
        "count": len(values),
        "min": min(values),
        "max": max(values),
        "avg": round(fmean(values), 3),
        "latest": readings[-1]["value"],
    }
